File: test/Coypu.Tests/When_interacting_with_the_browser/When_inspecting.cs
using System;
using NUnit.Framework;

package Coypu.Tests.When_interacting_with_the_browser
{
    public class When_inspecting : BrowserInteractionTests
    {
        protected void Queries_robustly(bool stubResult, Func<string, Options, bool> subject, Action<string, bool, DriverScope> stub)
        {
            Queries_robustly(stubResult, subject, stub, "Find me " + DateTime.Now.Ticks);
        }

        protected void Queries_robustly_reversing_result(bool stubResult, Func<string, Options, bool> subject, Action<string, bool, DriverScope> stub)
        {
            Queries_robustly_reversing_result(stubResult, subject, stub, "Find me " + DateTime.Now.Ticks);
        }

        protected void Queries_robustly<T>(bool stubResult, Func<T, Options, bool> subject, Action<T, bool, DriverScope> stub, T locator)
        {
            stub(locator, stubResult, browserSession);
            spyRobustWrapper.StubQueryResult(true, !stubResult);

            var individualTimeout = TimeSpan.FromMilliseconds(DateTime.UtcNow.Millisecond);

            var options = new Configuration {Timeout = individualTimeout};

            var actualImmediateResult = subject(locator, options);

            assertThat(actualImmediateResult, Is.EqualTo(!stubResult), "Result was not found robustly");

            RunQueryAndCheckTiming<bool>(individualTimeout);

            assertThat(queryResult, Is.EqualTo(stubResult));
        }

        protected void Queries_robustly_reversing_result<T>(bool stubResult, Func<T, Options, bool> subject, Action<T, bool, DriverScope> stub, T locator)
        {
            stub(locator, stubResult, browserSession);
            spyRobustWrapper.StubQueryResult(true, !stubResult);

            var actualImmediateResult = subject(locator,configuration);

            assertThat(actualImmediateResult, Is.EqualTo(!stubResult), "Result was not found robustly");

            RunQueryAndCheckTiming<bool>();

            assertThat(queryResult, Is.EqualTo(!stubResult));
        }
    }
}

File: src/test/java/Coypu.Tests/When_interacting_with_the_browser/When_inspecting.java
package Coypu.Tests.When_interacting_with_the_browser;

import Coypu.Options;
import Coypu.TimeSpan;
import org.junit.Before;

import java.util.UUID;

import static org.hamcrest.core.Is.is;
import static org.hamcrest.core.IsEqual.equalTo;
import static org.junit.Assert.assertThat;

public class When_inspecting extends BrowserInteractionTests {

    protected String locator;
    protected TimeSpan individualTimeout;
    protected Options options;

    @Before
    public void setUp() {
        locator = "Find me " + UUID.randomUUID();
        individualTimeout = TimeSpan.fromMilliseconds(6543);
        options = new Options();
        options.Timeout = individualTimeout;
    }

    protected void assertFoundRobustly(boolean stubResult, TimeSpan individualTimeout, boolean actualImmediateResult) {
        assertThat("Result was not found robustly", actualImmediateResult, is(equalTo(!stubResult)));

        Object queryResult = runQueryAndCheckTiming(individualTimeout);

        assertThat(queryResult, is(equalTo((Object) stubResult)));
    }

    protected void assertFoundRobustlyReversed(boolean stubResult, boolean actualImmediateResult) {
        assertThat("Result was not found robustly", actualImmediateResult, is(equalTo(!stubResult)));

        Object queryResult = runQueryAndCheckTiming();

        assertThat(queryResult, is(equalTo((Object) !stubResult)));
    }
}
